package/MDAnalysis/analysis/base.py

```diff
@@ -220,7 +220,165 @@ def __setstate__(self, state):
+            if not isinstance(analysis_object, AnalysisBase):
+                raise AttributeError(
+                    f"Analysis object {analysis_object} is "
+                    "not a child of `AnalysisBase`."
+                )
+
+        self._analysis_objects = analysis_objects
+
+    def run(
+        self,
+        start=None,
+        stop=None,
+        step=None,
+        frames=None,
+        verbose=None,
+        reset_timestep=True,
+    ):
+        """Perform the calculation
+
+        Parameters
+        ----------
+        start : int, optional
+            start frame of analysis
+        stop : int, optional
+            stop frame of analysis
+        step : int, optional
+            number of frames to skip between each analysed frame
+        frames : array_like, optional
+            array of integers or booleans to slice trajectory; `frames` can
+            only be used *instead* of `start`, `stop`, and `step`. Setting
+            *both* `frames` and at least one of `start`, `stop`, `step` to a
+            non-default value will raise a :exc:`ValueError`.
+        verbose : bool, optional
+            Turn on verbosity
+        reset_timestep : bool, optional
+            Reset the timestep object after for each ``analysis_object``.
+            Setting this to ``False`` can be useful if an ``analysis_object``
+            is performing a trajectory manipulation which is also useful for the
+            subsequent ``analysis_objects`` e.g. unwrapping of molecules.
+        """
+
+        # Ensure compatibility with API of version 0.15.0
+        if not hasattr(self, "_analysis_objects"):
+            self._analysis_objects = (self,)
+
+        logger.info("Choosing frames to analyze")
+        # if verbose unchanged, use class default
+        verbose = getattr(self, "_verbose", False) if verbose is None else verbose
+
+        logger.info("Starting preparation")
+
+        for analysis_object in self._analysis_objects:
+            analysis_object._setup_frames(
+                analysis_object._trajectory,
+                start=start,
+                stop=stop,
+                step=step,
+                frames=frames,
+            )
+            analysis_object._prepare()
+
+        logger.info(
+            f"Starting analysis loop over {self._analysis_objects[0].n_frames} "
+            "trajectory frames."
+        )
+
+        for i, ts in enumerate(
+            ProgressBar(self._analysis_objects[0]._sliced_trajectory, verbose=verbose)
+        ):
+            if reset_timestep:
+                ts_original = ts.copy()
+
+            for analysis_object in self._analysis_objects:
+                analysis_object._frame_index = i
+                analysis_object._ts = ts
+                analysis_object.frames[i] = ts.frame
+                analysis_object.times[i] = ts.time
+                analysis_object._single_frame()
+
+                if reset_timestep:
```

Comment:
Ts is never unassigned, so I'm not sure this is necessary here

Reply:
But, if an instance is changing the ts object we have to restore it from the stored one.